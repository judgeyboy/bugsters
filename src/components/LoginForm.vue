<template>
  <form v-on:submit="submit">
    <legend>Login</legend>
    <input type="text" placeholder="email" v-model="email">
    <input type="password" v-model="password">
    <input type="submit" value="Login">
  </form>
</template>

<script>

import axios from 'axios'

export default {
  name: 'LoginForm',
  data () {
    return {
      email: '',
      password: ''
    }
  },
  methods: {
    submit: function (submitEvent) {
      submitEvent.preventDefault()

      const loginDetails = {
        email: this.email,
        password: this.password
      }

      axios
        .post('https://bugsters-api.herokuapp.com/account/login', loginDetails, { headers: { 'Content-Type': 'application/json' } })
        .then(response => console.log(response))
        .catch(error => console.error(error))
    }
  }
}
</script>
