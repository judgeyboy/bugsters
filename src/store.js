import Vue from 'vue'
import Vuex from 'vuex'
import accountService from './services/account'

Vue.use(Vuex)

const store = new Vuex.Store({

  state: {
    loggedIn: false
  },

  mutations: {
    login (state) {
      state.loggedIn = true
    },
    logout (state) {
      state.loggedIn = false
    }
  },

  actions: {
    login ({ commit }, loginDetails) {
      accountService
        .logIn(loginDetails)
        .then(() => commit('login'))
        .catch(error => console.error(error))
    },
    logout ({ commit }) {
      commit('logout')
    }
  }

})

export default store
