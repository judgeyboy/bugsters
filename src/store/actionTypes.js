/*
  Account Actions
*/
export const ACCOUNT_LOGIN = 'ACCOUNT_LOGIN'
export const ACCOUNT_LOGOUT = 'ACCOUNT_LOGOUT'

/*
  Contact Actions
*/
export const CONTACT_SEND = 'CONTACT_SEND'
