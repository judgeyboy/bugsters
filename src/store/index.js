import Vue from 'vue'
import Vuex from 'vuex'
import accountModule from './modules/account'

Vue.use(Vuex)

const store = new Vuex.Store({
  modules: {
    accountModule
  }
})

export default store
