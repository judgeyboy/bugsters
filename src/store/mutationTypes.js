/*
  Account Mutations
*/
export const ACCOUNT_LOGIN = 'ACCOUNT_LOGIN'
export const ACCOUNT_LOGOUT = 'ACCOUNT_LOGOUT'

/*
  Contact Mutations
*/
export const CONTACT_SET_STATE = 'CONTACT_SET_STATE'
