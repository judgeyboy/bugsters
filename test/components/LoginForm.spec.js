import Vuex from 'vuex'
import VueRouter from 'vue-router'
import VeeValidate, { Validator } from 'vee-validate'
import { shallow, createLocalVue } from 'vue-test-utils'
import sinon from 'sinon'

import LoginForm from '@/components/LoginForm'
import { ACCOUNT_LOGIN } from '@/store/actionTypes'

const localVue = createLocalVue()

localVue.use(Vuex)
localVue.use(VeeValidate)
localVue.use(VueRouter)

const $router = new VueRouter()

describe('LoginForm component', () => {
  let store

  beforeEach(() => {
    store = new Vuex.Store({
      state: {},
      actions: {
        [ACCOUNT_LOGIN]: jest.fn()
      }
    })
  })

  it('should dispatch "login" with login details', async () => {
    const wrapper = shallow(LoginForm, {
      localVue,
      store,
      router: $router,
      provide: { $validator: new Validator() }
    })
    const spy = jest.spyOn(wrapper.vm.$store, 'dispatch')

    const loginDetails = {
      email: 'duchess@spy-agency.com',
      password: 'DangerZone!'
    }

    wrapper.setData(loginDetails)

    await wrapper.vm.onSubmit()

    expect(spy).toHaveBeenCalled()
    expect(spy).toHaveBeenCalledWith(ACCOUNT_LOGIN, loginDetails)
  })

  it('should not dispatch "login" if the form has errors', async () => {
    const wrapper = shallow(LoginForm, {
      localVue,
      store,
      provide: { $validator: new Validator() }
    })
    let spy = jest.spyOn(wrapper.vm.$store, 'dispatch')

    const invalidLoginDetails = {
      email: 'duchess@spy-agency.com',
      password: ''
    }

    wrapper.setData(invalidLoginDetails)

    await wrapper.vm.onSubmit()

    expect(spy).not.toHaveBeenCalled()
  })

  it('should redirect to a path after login success', async () => {
    const wrapper = shallow(LoginForm, {
      localVue,
      store,
      router: $router,
      provide: { $validator: new Validator() }
    })
    const spy = jest.spyOn(wrapper.vm.$router, 'push')

    const loginDetails = {
      email: 'duchess@spy-agency.com',
      password: 'DangerZone!'
    }
    wrapper.setData(loginDetails)
    wrapper.setProps({ returnUrl: '/about' })

    await wrapper.vm.onSubmit()

    expect(spy).toHaveBeenCalledWith({ path: '/about' })
  })

  it('should set isLoading to false and loginFailed to true if login fails', async () => {
    const wrapper = shallow(LoginForm, {
      localVue,
      store,
      router: $router,
      provide: { $validator: new Validator() }
    })
    const storeStub = sinon.stub(wrapper.vm.$store, 'dispatch').rejects()

    const loginDetails = {
      email: 'duchess@spy-agency.com',
      password: 'DangerZone!'
    }

    wrapper.setData(loginDetails)

    await wrapper.vm.onSubmit()

    wrapper.vm.$nextTick(() => {
      expect(wrapper.vm.loginFailed).toBe(true)
      expect(wrapper.vm.isLoading).toBe(false)
    })

    storeStub.restore()
  })

  /*
  * Skipping the snapshot tests because the shallow rendering does dont provide all the necessary elements
  */

  it.skip('snapshot', () => {
    const wrapper = shallow(LoginForm, {
      localVue,
      provide: {
        $validator: new Validator()
      }
    })
    const $html = wrapper.vm.$el.outerHTML
    expect($html).toMatchSnapshot()
  })

  it.skip('snapshot with login error', () => {
    const wrapper = shallow(LoginForm, {
      localVue,
      provide: {
        $validator: new Validator()
      }
    })

    wrapper.setData({ loginFailed: true })

    const $html = wrapper.vm.$el.outerHTML
    expect($html).toMatchSnapshot()
  })
})
