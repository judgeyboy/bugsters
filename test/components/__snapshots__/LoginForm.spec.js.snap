// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm component snapshot 1`] = `"<div class=\\"login-form\\"><div class=\\"container\\"><div class=\\"columns\\"><div class=\\"column col-6 col-sm-8 col-mx-auto\\"><h2 class=\\"section-title\\">Log In</h2> <!----> <!----> <form class=\\"form-horizontal\\"><div class=\\"form-group\\"><div class=\\"col-3 col-xs-12\\"><label for=\\"email\\" class=\\"form-label\\">Email</label></div> <div class=\\"col-9 col-xs-12\\"><input type=\\"email\\" name=\\"email\\" placeholder=\\"Email\\" class=\\"form-input\\" data-vv-id=\\"9\\" aria-required=\\"true\\" aria-invalid=\\"false\\"></div></div> <div class=\\"form-group\\"><div class=\\"col-3 col-xs-12\\"><label for=\\"password\\" class=\\"form-label\\">Password</label></div> <div class=\\"col-9 col-xs-12\\"><input type=\\"password\\" name=\\"password\\" placeholder=\\"password\\" class=\\"form-input\\" data-vv-id=\\"10\\" aria-required=\\"true\\" aria-invalid=\\"false\\"></div></div> <div class=\\"form-buttons\\"><button type=\\"submit\\" class=\\"btn btn-primary\\">Log In</button> <a href=\\"#/\\" class=\\"btn btn-link router-link-active\\">Cancel</a></div></form></div></div></div></div>"`;

exports[`LoginForm component snapshot with form error 1`] = `"<div class=\\"login-form\\"><div class=\\"container\\"><div class=\\"columns\\"><div class=\\"column col-6 col-sm-8 col-mx-auto\\"><h2 class=\\"section-title\\">Log In</h2> <!----> <div><span>Form contains errors, please check the red fields.</span></div> <form class=\\"form-horizontal\\"><div class=\\"form-group\\"><div class=\\"col-3 col-xs-12\\"><label for=\\"email\\" class=\\"form-label\\">Email</label></div> <div class=\\"col-9 col-xs-12\\"><input type=\\"email\\" name=\\"email\\" placeholder=\\"Email\\" class=\\"form-input\\" data-vv-id=\\"13\\" aria-required=\\"true\\" aria-invalid=\\"false\\"></div></div> <div class=\\"form-group\\"><div class=\\"col-3 col-xs-12\\"><label for=\\"password\\" class=\\"form-label\\">Password</label></div> <div class=\\"col-9 col-xs-12\\"><input type=\\"password\\" name=\\"password\\" placeholder=\\"password\\" class=\\"form-input\\" data-vv-id=\\"14\\" aria-required=\\"true\\" aria-invalid=\\"false\\"></div></div> <div class=\\"form-buttons\\"><button type=\\"submit\\" class=\\"btn btn-primary\\">Log In</button> <a href=\\"#/\\" class=\\"btn btn-link router-link-active\\">Cancel</a></div></form></div></div></div></div>"`;

exports[`LoginForm component snapshot with login error 1`] = `"<div class=\\"login-form\\"><div class=\\"container\\"><div class=\\"columns\\"><div class=\\"column col-6 col-sm-8 col-mx-auto\\"><h2 class=\\"section-title\\">Log In</h2> <div><span>Username or password does not exist!</span></div> <!----> <form class=\\"form-horizontal\\"><div class=\\"form-group\\"><div class=\\"col-3 col-xs-12\\"><label for=\\"email\\" class=\\"form-label\\">Email</label></div> <div class=\\"col-9 col-xs-12\\"><input type=\\"email\\" name=\\"email\\" placeholder=\\"Email\\" class=\\"form-input\\" data-vv-id=\\"11\\" aria-required=\\"true\\" aria-invalid=\\"false\\"></div></div> <div class=\\"form-group\\"><div class=\\"col-3 col-xs-12\\"><label for=\\"password\\" class=\\"form-label\\">Password</label></div> <div class=\\"col-9 col-xs-12\\"><input type=\\"password\\" name=\\"password\\" placeholder=\\"password\\" class=\\"form-input\\" data-vv-id=\\"12\\" aria-required=\\"true\\" aria-invalid=\\"false\\"></div></div> <div class=\\"form-buttons\\"><button type=\\"submit\\" class=\\"btn btn-primary\\">Log In</button> <a href=\\"#/\\" class=\\"btn btn-link router-link-active\\">Cancel</a></div></form></div></div></div></div>"`;
